### PR TITLE
Add support for Azure client certificate auth

### LINKED ISSUE #2761: Allow authentication using Service Principle with Certificate
**Describe the current behavior**
ASO currently only supports using a client secret for Service Principal-backed authentication.

**Describe the improvement**
Allow Service Principals to use a certificate for authentication like is described here: https://learn.microsoft.com/en-us/azure/developer/go/azure-sdk-authentication-service-principal?tabs=azure-cli

**Additional context**

### PR DESCRIPTION
**Pull request by super-harsh**
Closes #2761 

**What this PR does / why we need it**:

This PR adds support to handle client auth using Certificates

### CODE REVIEW

File: v2/cmd/controller/main.go [hunk 1]
```diff
@@ -179,6 +179,13 @@ func getDefaultAzureCredential(cfg config.Values, setupLog logr.Logger) (azcore.
 		if err != nil {
 			return nil, errors.Wrapf(err, "unable to get workload identity credential")
 		}
+	} else if cert := os.Getenv(config.ClientCertificateVar); cert != "" {
+		certPassword := os.Getenv(config.ClientCertificatePasswordVar)
+		credential, err = identity.NewClientCertificateCredential(cfg.TenantID, cfg.ClientID, []byte(cert), []byte(certPassword))
+		if err != nil {
+			setupLog.Error(err, "unable to get client certificate credential")
+			os.Exit(1)
```
Comment (matthchr) [inline v2/cmd/controller/main.go:187]: This should return an error, not `os.Exit`.

Let the caller decide to exit or not, like with the other errors here.

File: v2/test/single_operator_multitenant_test.go [hunk 2]
```diff
@@ -106,21 +138,47 @@ func Test_Multitenant_SingleOperator_PerResourceCredential(t *testing.T) {
...
 
-	secretData[config.ClientIDVar] = []byte(clientID)
-	secretData[config.ClientSecretVar] = []byte(clientSecret)
+	secret.Data[config.ClientIDVar] = []byte(clientID)
+	secret.Data[config.ClientCertificateVar] = []byte(clientCert)
+
+	return secret, nil
+}
+
+func newCredentialSecret(subscriptionID, tenantID, name, namespaceName string) *v1.Secret {
```
Comment (matthchr) [inline v2/test/single_operator_multitenant_test.go:179]: minor: IMO just call the `namespaceName` parameter `namespace`

File: v2/test/single_operator_multitenant_test.go [hunk 3]
```diff
@@ -106,21 +138,47 @@ func Test_Multitenant_SingleOperator_PerResourceCredential(t *testing.T) {
...
+		return nil, errors.Errorf("required environment variable %q was not supplied", AzureClientIDMultitenantVar)
+	}
+
+	secret.Data[config.ClientIDVar] = []byte(clientID)
+	secret.Data[config.ClientSecretVar] = []byte(clientSecret)
+
+	return secret, nil
+}
+
+func newClientCertificateCredential(subscriptionID, tenantID, name, namespaceName string) (*v1.Secret, error) {
```
Comment (matthchr) [inline v2/test/single_operator_multitenant_test.go:160]: minor: `namespaceName` -> `namespace`

**Review by matthchr — COMMENTED** [review]
Basically LGTM, left a few small comments. Happy to approve once those are resolved.

File: v2/charts/azure-service-operator/values.yaml [hunk 4]
```diff
@@ -13,6 +13,12 @@ azureClientID: ""
 # This is required when using Service Principal authentication.
 azureClientSecret: ""
 
+#  A PEM or PKCS12 certificate file including the private key for Azure Credential Authentication. The certificate must not be password-protected.
```
[the remark this answers is not in the document]
Comment (super-harsh) [reply v2/charts/azure-service-operator/values.yaml:16]: We do allow password protected certs, updating the comment.

File: docs/hugo/content/introduction/authentication.md [hunk 5]
```diff
@@ -43,6 +45,49 @@ stringData:
...
+
+For more information about Service Principals, see [creating an Azure Service Principal using the Azure CLI](https://learn.microsoft.com/cli/azure/create-an-azure-service-principal-azure-cli#certificate-based-authentication).
+The `AZURE_CLIENT_ID` is sometimes also called the App ID. The `AZURE_CLIENT_CERTIFICATE` is the "certificate" returned by the command in the previously linked documentation.
+
+Use the following Bash script to set the environment variables for the `aso-controller-settings` secret:
+```bash
+export AZURE_CLIENT_ID="00000000-0000-0000-0000-00000000000"          # The client ID (sometimes called App Id) of the Service Principal.
+export AZURE_SUBSCRIPTION_ID="00000000-0000-0000-0000-00000000000"    # The Azure Subscription ID the identity is in.
+export AZURE_TENANT_ID="00000000-0000-0000-0000-00000000000"          # The Azure AAD Tenant the identity/subscription is associated with.
+export AZURE_CLIENT_CERTIFICATE="-----BEGIN PRIVATE KEY-----
```
Comment (matthchr) [inline docs/hugo/content/introduction/authentication.md:64]: minor: does this export command actually work? Not sure how multiline exports work.

**Review by matthchr — APPROVED** [review]

File: docs/hugo/content/introduction/authentication.md (hunk 5 above)
Comment (super-harsh) [reply docs/hugo/content/introduction/authentication.md:64]: yes, it does.. will have to do something like `export AZURE_CLIENT_CERTIFICATE=`cat certFile.pem`. Also, we have a livetest for this.

**Review by theunrepentantgeek — APPROVED** [review]

**Pull request merged**